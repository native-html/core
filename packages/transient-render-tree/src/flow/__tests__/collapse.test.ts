import { TNode } from '../../tree/TNode';
import { collapse } from '../collapse';
import { hoist } from '../hoist';
import {
  deeplyNestedSource,
  href,
  imgSrc,
  nestedHyperlinksSource,
  recursiveHoisting,
  rfc002Source,
  secondaryHref
} from './shared';
import { translateTreeTest } from './utils';

async function makeTTree(html: string): Promise<TNode> {
  return collapse(hoist(await translateTreeTest(html)));
}

describe('collapse function', () => {
  it('should collapse a tree such as specified in RFC002 example', async () => {
    const ttree = await makeTTree(rfc002Source);
    expect(ttree).toMatchObject({
      type: 'block',
      isAnchor: false,
      children: [
        {
          type: 'phrasing',
          isAnchor: true,
          href,
          children: [
            {
              type: 'text',
              data: 'This is '
            },
            {
              type: 'text',
              data: 'phrasing content',
              tagName: 'span'
            }
          ]
        },
        {
          type: 'block',
          isAnchor: true,
          tagName: 'img',
          attributes: {
            src: imgSrc
          }
        },
        {
          type: 'phrasing',
          isAnchor: true,
          href,
          children: [
            {
              type: 'text',
              data: 'and this is '
            },
            {
              type: 'text',
              tagName: 'strong',
              data: 'too'
            },
            {
              type: 'text',
              data: '.'
            }
          ]
        }
      ]
    });
    expect(ttree).toMatchSnapshot();
  });
  it('should handle nested anchors', async () => {
    const ttree = await makeTTree(nestedHyperlinksSource);
    expect(ttree).toMatchObject({
      type: 'phrasing',
      isAnchor: true,
      href: href,
      children: [
        {
          type: 'text',
          data: 'This is '
        },
        {
          type: 'text',
          tagName: 'span',
          data: 'phrasing content'
        },
        {
          type: 'text',
          data: ' '
        },
        {
          type: 'phrasing',
          isAnchor: true,
          href: secondaryHref,
          children: [
            {
              type: 'text',
              data: 'and this is '
            },
            {
              type: 'text',
              tagName: 'strong',
              data: 'too'
            }
          ]
        }
      ]
    });
    expect(ttree).toMatchSnapshot();
  });
  it('should handle deeply nested HTML', async () => {
    const ttree = await makeTTree(deeplyNestedSource);
    expect(ttree).toMatchObject({
      type: 'block',
      tagName: 'article',
      children: [
        {
          type: 'phrasing',
          tagName: null,
          children: [
            {
              type: 'text',
              data: 'First implicit paragraph.'
            }
          ]
        },
        {
          type: 'block',
          tagName: 'div',
          children: [
            {
              type: 'phrasing',
              tagName: null,
              children: [
                {
                  type: 'text',
                  data: 'Second implicit paragraph'
                }
              ]
            },
            {
              type: 'block',
              tagName: 'p',
              children: [
                {
                  type: 'phrasing',
                  children: [
                    {
                      type: 'text',
                      data: 'A new paragraph.'
                    },
                    {
                      type: 'text',
                      tagName: 'span',
                      data: ' And a span within'
                    }
                  ]
                },
                {
                  type: 'block',
                  tagName: 'img'
                }
              ]
            }
          ]
        }
      ]
    });
    expect(ttree).toMatchSnapshot();
  });
  it('should hoist blocks recursively', async () => {
    const ttree = await makeTTree(recursiveHoisting);
    expect(ttree).toMatchObject({
      type: 'block',
      tagName: 'span',
      children: [
        {
          type: 'phrasing',
          tagName: null,
          children: [
            {
              type: 'text',
              data: 'Line1',
              tagName: null
            }
          ]
        },
        {
          type: 'block',
          tagName: 'strong',
          children: [
            {
              type: 'phrasing',
              tagName: null,
              children: [
                {
                  type: 'text',
                  data: 'Line2',
                  tagName: null
                },
                {
                  type: 'text',
                  data: ' Line3',
                  tagName: 'span'
                }
              ]
            },
            {
              type: 'block',
              tagName: 'img',
              attributes: { src: imgSrc }
            }
          ]
        }
      ]
    });
    expect(ttree).toMatchSnapshot();
  });
  it('should handle body tag', async () => {
    const html = `<body><div>
    <span>Hello world!</span>
</div></body>`;
    const ttree = await makeTTree(html);
    expect(ttree).toMatchObject({
      type: 'block',
      tagName: 'body',
      children: [
        {
          type: 'block',
          tagName: 'div',
          children: [
            {
              type: 'phrasing',
              tagName: null,
              children: [
                {
                  type: 'text',
                  tagName: 'span',
                  data: 'Hello world!'
                }
              ]
            }
          ]
        }
      ]
    });
  });
  it('should handle direct style inheritance', async () => {
    const ttree = await makeTTree(
      '<div style="font-size: 18px;border-width: 20px;"><span style="color: red;">This is nice!</span></div>'
    );
    expect(ttree).toMatchObject({
      type: 'block',
      isAnchor: false,
      attributes: {},
      tagName: 'div',
      styles: {
        nativeTextFlow: { fontSize: 18 },
        nativeBlockRet: {
          borderTopWidth: 20,
          borderRightWidth: 20,
          borderBottomWidth: 20,
          borderLeftWidth: 20
        }
      },
      children: [
        {
          type: 'phrasing',
          isAnchor: false,
          attributes: {},
          styles: {
            nativeTextFlow: { fontSize: 18 },
            nativeBlockRet: {}
          },
          children: [
            {
              type: 'text',
              tagName: 'span',
              data: 'This is nice!',
              styles: {
                nativeTextFlow: { color: 'red' }
              }
            }
          ]
        }
      ]
    });
    expect(ttree).toMatchSnapshot();
  });
  it('should handle indirect style inheritance', async () => {
    const ttree = await makeTTree(
      '<div style="font-size: 18px;"><div style="border-width: 20px;">This is great!</div></div>'
    );
    expect(ttree).toMatchObject({
      type: 'block',
      isAnchor: false,
      attributes: {},
      tagName: 'div',
      styles: {
        nativeTextFlow: { fontSize: 18 }
      },
      children: [
        {
          type: 'block',
          tagName: 'div',
          styles: {
            nativeTextFlow: { fontSize: 18 },
            nativeBlockRet: {
              borderTopWidth: 20,
              borderRightWidth: 20,
              borderBottomWidth: 20,
              borderLeftWidth: 20
            }
          },
          children: [
            {
              type: 'phrasing',
              tagName: null,
              styles: {
                nativeTextFlow: { fontSize: 18 }
              },
              children: [
                {
                  type: 'text',
                  data: 'This is great!',
                  tagName: null
                }
              ]
            }
          ]
        }
      ]
    });
    expect(ttree).toMatchSnapshot();
  });
});
