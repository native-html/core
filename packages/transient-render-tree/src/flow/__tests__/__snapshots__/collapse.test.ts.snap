// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`collapse function should collapse a tree such as specified in RFC002 example 1`] = `
TBlock {
  "attributes": Object {
    "href": "https://domain.com",
  },
  "children": Array [
    TPhrasingAnchor {
      "attributes": Object {},
      "children": Array [
        TText {
          "attributes": Object {},
          "children": Array [],
          "data": "This is ",
          "hasWhiteSpaceCollapsingEnabled": true,
          "isAnchor": false,
          "parentStyles": TStyles {
            "nativeBlockFlow": Object {},
            "nativeBlockRet": Object {},
            "nativeTextFlow": Object {},
            "nativeTextRet": Object {},
            "webTextFlow": Object {},
          },
          "styles": TStyles {
            "nativeBlockFlow": Object {},
            "nativeBlockRet": Object {},
            "nativeTextFlow": Object {},
            "nativeTextRet": Object {},
            "webTextFlow": Object {},
          },
          "tagName": null,
          "type": "text",
        },
        TText {
          "attributes": Object {},
          "children": Array [],
          "data": "phrasing content",
          "hasWhiteSpaceCollapsingEnabled": true,
          "isAnchor": false,
          "parentStyles": TStyles {
            "nativeBlockFlow": Object {},
            "nativeBlockRet": Object {},
            "nativeTextFlow": Object {},
            "nativeTextRet": Object {},
            "webTextFlow": Object {},
          },
          "styles": TStyles {
            "nativeBlockFlow": Object {},
            "nativeBlockRet": Object {},
            "nativeTextFlow": Object {},
            "nativeTextRet": Object {},
            "webTextFlow": Object {},
          },
          "tagName": "span",
          "type": "text",
        },
      ],
      "hasWhiteSpaceCollapsingEnabled": true,
      "href": "https://domain.com",
      "isAnchor": true,
      "parentStyles": null,
      "styles": TStyles {
        "nativeBlockFlow": Object {},
        "nativeBlockRet": Object {},
        "nativeTextFlow": Object {},
        "nativeTextRet": Object {},
        "webTextFlow": Object {},
      },
      "tagName": null,
      "type": "phrasing",
    },
    TBlockAnchor {
      "attributes": Object {
        "src": "https://domain.com/logo.jpg",
      },
      "children": Array [],
      "hasWhiteSpaceCollapsingEnabled": true,
      "href": "https://domain.com",
      "isAnchor": true,
      "parentStyles": null,
      "styles": TStyles {
        "nativeBlockFlow": Object {},
        "nativeBlockRet": Object {},
        "nativeTextFlow": Object {},
        "nativeTextRet": Object {},
        "webTextFlow": Object {},
      },
      "tagName": "img",
      "type": "block",
    },
    TPhrasingAnchor {
      "attributes": Object {},
      "children": Array [
        TText {
          "attributes": Object {},
          "children": Array [],
          "data": "and this is ",
          "hasWhiteSpaceCollapsingEnabled": true,
          "isAnchor": false,
          "parentStyles": TStyles {
            "nativeBlockFlow": Object {},
            "nativeBlockRet": Object {},
            "nativeTextFlow": Object {},
            "nativeTextRet": Object {},
            "webTextFlow": Object {},
          },
          "styles": TStyles {
            "nativeBlockFlow": Object {},
            "nativeBlockRet": Object {},
            "nativeTextFlow": Object {},
            "nativeTextRet": Object {},
            "webTextFlow": Object {},
          },
          "tagName": null,
          "type": "text",
        },
        TText {
          "attributes": Object {},
          "children": Array [],
          "data": "too",
          "hasWhiteSpaceCollapsingEnabled": true,
          "isAnchor": false,
          "parentStyles": TStyles {
            "nativeBlockFlow": Object {},
            "nativeBlockRet": Object {},
            "nativeTextFlow": Object {},
            "nativeTextRet": Object {},
            "webTextFlow": Object {},
          },
          "styles": TStyles {
            "nativeBlockFlow": Object {},
            "nativeBlockRet": Object {},
            "nativeTextFlow": Object {},
            "nativeTextRet": Object {},
            "webTextFlow": Object {},
          },
          "tagName": "strong",
          "type": "text",
        },
        TText {
          "attributes": Object {},
          "children": Array [],
          "data": ".",
          "hasWhiteSpaceCollapsingEnabled": true,
          "isAnchor": false,
          "parentStyles": TStyles {
            "nativeBlockFlow": Object {},
            "nativeBlockRet": Object {},
            "nativeTextFlow": Object {},
            "nativeTextRet": Object {},
            "webTextFlow": Object {},
          },
          "styles": TStyles {
            "nativeBlockFlow": Object {},
            "nativeBlockRet": Object {},
            "nativeTextFlow": Object {},
            "nativeTextRet": Object {},
            "webTextFlow": Object {},
          },
          "tagName": null,
          "type": "text",
        },
      ],
      "hasWhiteSpaceCollapsingEnabled": true,
      "href": "https://domain.com",
      "isAnchor": true,
      "parentStyles": null,
      "styles": TStyles {
        "nativeBlockFlow": Object {},
        "nativeBlockRet": Object {},
        "nativeTextFlow": Object {},
        "nativeTextRet": Object {},
        "webTextFlow": Object {},
      },
      "tagName": null,
      "type": "phrasing",
    },
  ],
  "hasWhiteSpaceCollapsingEnabled": true,
  "isAnchor": false,
  "parentStyles": null,
  "styles": TStyles {
    "nativeBlockFlow": Object {},
    "nativeBlockRet": Object {},
    "nativeTextFlow": Object {},
    "nativeTextRet": Object {},
    "webTextFlow": Object {},
  },
  "tagName": "a",
  "type": "block",
}
`;

exports[`collapse function should handle deeply nested HTML 1`] = `
TBlock {
  "attributes": Object {},
  "children": Array [
    TPhrasing {
      "attributes": Object {},
      "children": Array [
        TText {
          "attributes": Object {},
          "children": Array [],
          "data": "First implicit paragraph.",
          "hasWhiteSpaceCollapsingEnabled": true,
          "isAnchor": false,
          "parentStyles": TStyles {
            "nativeBlockFlow": Object {},
            "nativeBlockRet": Object {},
            "nativeTextFlow": Object {},
            "nativeTextRet": Object {},
            "webTextFlow": Object {},
          },
          "styles": TStyles {
            "nativeBlockFlow": Object {},
            "nativeBlockRet": Object {},
            "nativeTextFlow": Object {},
            "nativeTextRet": Object {},
            "webTextFlow": Object {},
          },
          "tagName": null,
          "type": "text",
        },
      ],
      "hasWhiteSpaceCollapsingEnabled": true,
      "isAnchor": false,
      "parentStyles": TStyles {
        "nativeBlockFlow": Object {},
        "nativeBlockRet": Object {},
        "nativeTextFlow": Object {},
        "nativeTextRet": Object {},
        "webTextFlow": Object {},
      },
      "styles": TStyles {
        "nativeBlockFlow": Object {},
        "nativeBlockRet": Object {},
        "nativeTextFlow": Object {},
        "nativeTextRet": Object {},
        "webTextFlow": Object {},
      },
      "tagName": null,
      "type": "phrasing",
    },
    TBlock {
      "attributes": Object {},
      "children": Array [
        TPhrasing {
          "attributes": Object {},
          "children": Array [
            TText {
              "attributes": Object {},
              "children": Array [],
              "data": "Second implicit paragraph",
              "hasWhiteSpaceCollapsingEnabled": true,
              "isAnchor": false,
              "parentStyles": TStyles {
                "nativeBlockFlow": Object {},
                "nativeBlockRet": Object {},
                "nativeTextFlow": Object {},
                "nativeTextRet": Object {},
                "webTextFlow": Object {},
              },
              "styles": TStyles {
                "nativeBlockFlow": Object {},
                "nativeBlockRet": Object {},
                "nativeTextFlow": Object {},
                "nativeTextRet": Object {},
                "webTextFlow": Object {},
              },
              "tagName": null,
              "type": "text",
            },
          ],
          "hasWhiteSpaceCollapsingEnabled": true,
          "isAnchor": false,
          "parentStyles": TStyles {
            "nativeBlockFlow": Object {},
            "nativeBlockRet": Object {},
            "nativeTextFlow": Object {},
            "nativeTextRet": Object {},
            "webTextFlow": Object {},
          },
          "styles": TStyles {
            "nativeBlockFlow": Object {},
            "nativeBlockRet": Object {},
            "nativeTextFlow": Object {},
            "nativeTextRet": Object {},
            "webTextFlow": Object {},
          },
          "tagName": null,
          "type": "phrasing",
        },
        TBlock {
          "attributes": Object {},
          "children": Array [
            TPhrasing {
              "attributes": Object {},
              "children": Array [
                TText {
                  "attributes": Object {},
                  "children": Array [],
                  "data": "A new paragraph.",
                  "hasWhiteSpaceCollapsingEnabled": true,
                  "isAnchor": false,
                  "parentStyles": TStyles {
                    "nativeBlockFlow": Object {},
                    "nativeBlockRet": Object {},
                    "nativeTextFlow": Object {},
                    "nativeTextRet": Object {},
                    "webTextFlow": Object {},
                  },
                  "styles": TStyles {
                    "nativeBlockFlow": Object {},
                    "nativeBlockRet": Object {},
                    "nativeTextFlow": Object {},
                    "nativeTextRet": Object {},
                    "webTextFlow": Object {},
                  },
                  "tagName": null,
                  "type": "text",
                },
                TText {
                  "attributes": Object {},
                  "children": Array [],
                  "data": " And a span within",
                  "hasWhiteSpaceCollapsingEnabled": true,
                  "isAnchor": false,
                  "parentStyles": TStyles {
                    "nativeBlockFlow": Object {},
                    "nativeBlockRet": Object {},
                    "nativeTextFlow": Object {},
                    "nativeTextRet": Object {},
                    "webTextFlow": Object {},
                  },
                  "styles": TStyles {
                    "nativeBlockFlow": Object {},
                    "nativeBlockRet": Object {},
                    "nativeTextFlow": Object {},
                    "nativeTextRet": Object {},
                    "webTextFlow": Object {},
                  },
                  "tagName": "span",
                  "type": "text",
                },
              ],
              "hasWhiteSpaceCollapsingEnabled": true,
              "isAnchor": false,
              "parentStyles": TStyles {
                "nativeBlockFlow": Object {},
                "nativeBlockRet": Object {},
                "nativeTextFlow": Object {},
                "nativeTextRet": Object {},
                "webTextFlow": Object {},
              },
              "styles": TStyles {
                "nativeBlockFlow": Object {},
                "nativeBlockRet": Object {},
                "nativeTextFlow": Object {},
                "nativeTextRet": Object {},
                "webTextFlow": Object {},
              },
              "tagName": null,
              "type": "phrasing",
            },
            TBlock {
              "attributes": Object {
                "src": "https://domain.com/logo.jpg",
              },
              "children": Array [],
              "hasWhiteSpaceCollapsingEnabled": true,
              "isAnchor": false,
              "parentStyles": TStyles {
                "nativeBlockFlow": Object {},
                "nativeBlockRet": Object {},
                "nativeTextFlow": Object {},
                "nativeTextRet": Object {},
                "webTextFlow": Object {},
              },
              "styles": TStyles {
                "nativeBlockFlow": Object {},
                "nativeBlockRet": Object {},
                "nativeTextFlow": Object {},
                "nativeTextRet": Object {},
                "webTextFlow": Object {},
              },
              "tagName": "img",
              "type": "block",
            },
          ],
          "hasWhiteSpaceCollapsingEnabled": true,
          "isAnchor": false,
          "parentStyles": TStyles {
            "nativeBlockFlow": Object {},
            "nativeBlockRet": Object {},
            "nativeTextFlow": Object {},
            "nativeTextRet": Object {},
            "webTextFlow": Object {},
          },
          "styles": TStyles {
            "nativeBlockFlow": Object {},
            "nativeBlockRet": Object {},
            "nativeTextFlow": Object {},
            "nativeTextRet": Object {},
            "webTextFlow": Object {},
          },
          "tagName": "p",
          "type": "block",
        },
      ],
      "hasWhiteSpaceCollapsingEnabled": true,
      "isAnchor": false,
      "parentStyles": TStyles {
        "nativeBlockFlow": Object {},
        "nativeBlockRet": Object {},
        "nativeTextFlow": Object {},
        "nativeTextRet": Object {},
        "webTextFlow": Object {},
      },
      "styles": TStyles {
        "nativeBlockFlow": Object {},
        "nativeBlockRet": Object {},
        "nativeTextFlow": Object {},
        "nativeTextRet": Object {},
        "webTextFlow": Object {},
      },
      "tagName": "div",
      "type": "block",
    },
  ],
  "hasWhiteSpaceCollapsingEnabled": true,
  "isAnchor": false,
  "parentStyles": null,
  "styles": TStyles {
    "nativeBlockFlow": Object {},
    "nativeBlockRet": Object {},
    "nativeTextFlow": Object {},
    "nativeTextRet": Object {},
    "webTextFlow": Object {},
  },
  "tagName": "article",
  "type": "block",
}
`;

exports[`collapse function should handle direct style inheritance 1`] = `
TBlock {
  "attributes": Object {},
  "children": Array [
    TPhrasing {
      "attributes": Object {},
      "children": Array [
        TText {
          "attributes": Object {},
          "children": Array [],
          "data": "This is nice!",
          "hasWhiteSpaceCollapsingEnabled": true,
          "isAnchor": false,
          "parentStyles": TStyles {
            "nativeBlockFlow": Object {},
            "nativeBlockRet": Object {
              "borderBottomWidth": 20,
              "borderLeftWidth": 20,
              "borderRightWidth": 20,
              "borderTopWidth": 20,
            },
            "nativeTextFlow": Object {},
            "nativeTextRet": Object {},
            "webTextFlow": Object {},
          },
          "styles": TStyles {
            "nativeBlockFlow": Object {},
            "nativeBlockRet": Object {},
            "nativeTextFlow": Object {},
            "nativeTextRet": Object {},
            "webTextFlow": Object {},
          },
          "tagName": "span",
          "type": "text",
        },
      ],
      "hasWhiteSpaceCollapsingEnabled": true,
      "isAnchor": false,
      "parentStyles": TStyles {
        "nativeBlockFlow": Object {},
        "nativeBlockRet": Object {
          "borderBottomWidth": 20,
          "borderLeftWidth": 20,
          "borderRightWidth": 20,
          "borderTopWidth": 20,
        },
        "nativeTextFlow": Object {},
        "nativeTextRet": Object {},
        "webTextFlow": Object {},
      },
      "styles": TStyles {
        "nativeBlockFlow": Object {},
        "nativeBlockRet": Object {},
        "nativeTextFlow": Object {},
        "nativeTextRet": Object {},
        "webTextFlow": Object {},
      },
      "tagName": null,
      "type": "phrasing",
    },
  ],
  "hasWhiteSpaceCollapsingEnabled": true,
  "isAnchor": false,
  "parentStyles": null,
  "styles": TStyles {
    "nativeBlockFlow": Object {},
    "nativeBlockRet": Object {
      "borderBottomWidth": 20,
      "borderLeftWidth": 20,
      "borderRightWidth": 20,
      "borderTopWidth": 20,
    },
    "nativeTextFlow": Object {},
    "nativeTextRet": Object {},
    "webTextFlow": Object {},
  },
  "tagName": "div",
  "type": "block",
}
`;

exports[`collapse function should handle indirect style inheritance 1`] = `
TBlock {
  "attributes": Object {},
  "children": Array [
    TBlock {
      "attributes": Object {},
      "children": Array [
        TPhrasing {
          "attributes": Object {},
          "children": Array [
            TText {
              "attributes": Object {},
              "children": Array [],
              "data": "This is great!",
              "hasWhiteSpaceCollapsingEnabled": true,
              "isAnchor": false,
              "parentStyles": TStyles {
                "nativeBlockFlow": Object {},
                "nativeBlockRet": Object {
                  "borderBottomWidth": 20,
                  "borderLeftWidth": 20,
                  "borderRightWidth": 20,
                  "borderTopWidth": 20,
                },
                "nativeTextFlow": Object {},
                "nativeTextRet": Object {},
                "webTextFlow": Object {},
              },
              "styles": TStyles {
                "nativeBlockFlow": Object {},
                "nativeBlockRet": Object {},
                "nativeTextFlow": Object {},
                "nativeTextRet": Object {},
                "webTextFlow": Object {},
              },
              "tagName": null,
              "type": "text",
            },
          ],
          "hasWhiteSpaceCollapsingEnabled": true,
          "isAnchor": false,
          "parentStyles": TStyles {
            "nativeBlockFlow": Object {},
            "nativeBlockRet": Object {
              "borderBottomWidth": 20,
              "borderLeftWidth": 20,
              "borderRightWidth": 20,
              "borderTopWidth": 20,
            },
            "nativeTextFlow": Object {},
            "nativeTextRet": Object {},
            "webTextFlow": Object {},
          },
          "styles": TStyles {
            "nativeBlockFlow": Object {},
            "nativeBlockRet": Object {},
            "nativeTextFlow": Object {},
            "nativeTextRet": Object {},
            "webTextFlow": Object {},
          },
          "tagName": null,
          "type": "phrasing",
        },
      ],
      "hasWhiteSpaceCollapsingEnabled": true,
      "isAnchor": false,
      "parentStyles": TStyles {
        "nativeBlockFlow": Object {},
        "nativeBlockRet": Object {},
        "nativeTextFlow": Object {},
        "nativeTextRet": Object {},
        "webTextFlow": Object {},
      },
      "styles": TStyles {
        "nativeBlockFlow": Object {},
        "nativeBlockRet": Object {
          "borderBottomWidth": 20,
          "borderLeftWidth": 20,
          "borderRightWidth": 20,
          "borderTopWidth": 20,
        },
        "nativeTextFlow": Object {},
        "nativeTextRet": Object {},
        "webTextFlow": Object {},
      },
      "tagName": "div",
      "type": "block",
    },
  ],
  "hasWhiteSpaceCollapsingEnabled": true,
  "isAnchor": false,
  "parentStyles": null,
  "styles": TStyles {
    "nativeBlockFlow": Object {},
    "nativeBlockRet": Object {},
    "nativeTextFlow": Object {},
    "nativeTextRet": Object {},
    "webTextFlow": Object {},
  },
  "tagName": "div",
  "type": "block",
}
`;

exports[`collapse function should handle nested anchors 1`] = `
TPhrasingAnchor {
  "attributes": Object {
    "href": "https://domain.com",
  },
  "children": Array [
    TText {
      "attributes": Object {},
      "children": Array [],
      "data": "This is ",
      "hasWhiteSpaceCollapsingEnabled": true,
      "isAnchor": false,
      "parentStyles": TStyles {
        "nativeBlockFlow": Object {},
        "nativeBlockRet": Object {},
        "nativeTextFlow": Object {},
        "nativeTextRet": Object {},
        "webTextFlow": Object {},
      },
      "styles": TStyles {
        "nativeBlockFlow": Object {},
        "nativeBlockRet": Object {},
        "nativeTextFlow": Object {},
        "nativeTextRet": Object {},
        "webTextFlow": Object {},
      },
      "tagName": null,
      "type": "text",
    },
    TText {
      "attributes": Object {},
      "children": Array [],
      "data": "phrasing content",
      "hasWhiteSpaceCollapsingEnabled": true,
      "isAnchor": false,
      "parentStyles": TStyles {
        "nativeBlockFlow": Object {},
        "nativeBlockRet": Object {},
        "nativeTextFlow": Object {},
        "nativeTextRet": Object {},
        "webTextFlow": Object {},
      },
      "styles": TStyles {
        "nativeBlockFlow": Object {},
        "nativeBlockRet": Object {},
        "nativeTextFlow": Object {},
        "nativeTextRet": Object {},
        "webTextFlow": Object {},
      },
      "tagName": "span",
      "type": "text",
    },
    TText {
      "attributes": Object {},
      "children": Array [],
      "data": " ",
      "hasWhiteSpaceCollapsingEnabled": true,
      "isAnchor": false,
      "parentStyles": TStyles {
        "nativeBlockFlow": Object {},
        "nativeBlockRet": Object {},
        "nativeTextFlow": Object {},
        "nativeTextRet": Object {},
        "webTextFlow": Object {},
      },
      "styles": TStyles {
        "nativeBlockFlow": Object {},
        "nativeBlockRet": Object {},
        "nativeTextFlow": Object {},
        "nativeTextRet": Object {},
        "webTextFlow": Object {},
      },
      "tagName": null,
      "type": "text",
    },
    TPhrasingAnchor {
      "attributes": Object {
        "href": "https://sub.domain.com",
      },
      "children": Array [
        TText {
          "attributes": Object {},
          "children": Array [],
          "data": "and this is ",
          "hasWhiteSpaceCollapsingEnabled": true,
          "isAnchor": false,
          "parentStyles": TStyles {
            "nativeBlockFlow": Object {},
            "nativeBlockRet": Object {},
            "nativeTextFlow": Object {},
            "nativeTextRet": Object {},
            "webTextFlow": Object {},
          },
          "styles": TStyles {
            "nativeBlockFlow": Object {},
            "nativeBlockRet": Object {},
            "nativeTextFlow": Object {},
            "nativeTextRet": Object {},
            "webTextFlow": Object {},
          },
          "tagName": null,
          "type": "text",
        },
        TText {
          "attributes": Object {},
          "children": Array [],
          "data": "too",
          "hasWhiteSpaceCollapsingEnabled": true,
          "isAnchor": false,
          "parentStyles": TStyles {
            "nativeBlockFlow": Object {},
            "nativeBlockRet": Object {},
            "nativeTextFlow": Object {},
            "nativeTextRet": Object {},
            "webTextFlow": Object {},
          },
          "styles": TStyles {
            "nativeBlockFlow": Object {},
            "nativeBlockRet": Object {},
            "nativeTextFlow": Object {},
            "nativeTextRet": Object {},
            "webTextFlow": Object {},
          },
          "tagName": "strong",
          "type": "text",
        },
      ],
      "hasWhiteSpaceCollapsingEnabled": true,
      "href": "https://sub.domain.com",
      "isAnchor": true,
      "parentStyles": TStyles {
        "nativeBlockFlow": Object {},
        "nativeBlockRet": Object {},
        "nativeTextFlow": Object {},
        "nativeTextRet": Object {},
        "webTextFlow": Object {},
      },
      "styles": TStyles {
        "nativeBlockFlow": Object {},
        "nativeBlockRet": Object {},
        "nativeTextFlow": Object {},
        "nativeTextRet": Object {},
        "webTextFlow": Object {},
      },
      "tagName": "a",
      "type": "phrasing",
    },
  ],
  "hasWhiteSpaceCollapsingEnabled": true,
  "href": "https://domain.com",
  "isAnchor": true,
  "parentStyles": null,
  "styles": TStyles {
    "nativeBlockFlow": Object {},
    "nativeBlockRet": Object {},
    "nativeTextFlow": Object {},
    "nativeTextRet": Object {},
    "webTextFlow": Object {},
  },
  "tagName": "a",
  "type": "phrasing",
}
`;

exports[`collapse function should hoist blocks recursively 1`] = `
TBlock {
  "attributes": Object {},
  "children": Array [
    TPhrasing {
      "attributes": Object {},
      "children": Array [
        TText {
          "attributes": Object {},
          "children": Array [],
          "data": "Line1",
          "hasWhiteSpaceCollapsingEnabled": true,
          "isAnchor": false,
          "parentStyles": TStyles {
            "nativeBlockFlow": Object {},
            "nativeBlockRet": Object {},
            "nativeTextFlow": Object {},
            "nativeTextRet": Object {},
            "webTextFlow": Object {},
          },
          "styles": TStyles {
            "nativeBlockFlow": Object {},
            "nativeBlockRet": Object {},
            "nativeTextFlow": Object {},
            "nativeTextRet": Object {},
            "webTextFlow": Object {},
          },
          "tagName": null,
          "type": "text",
        },
      ],
      "hasWhiteSpaceCollapsingEnabled": true,
      "isAnchor": false,
      "parentStyles": null,
      "styles": TStyles {
        "nativeBlockFlow": Object {},
        "nativeBlockRet": Object {},
        "nativeTextFlow": Object {},
        "nativeTextRet": Object {},
        "webTextFlow": Object {},
      },
      "tagName": null,
      "type": "phrasing",
    },
    TBlock {
      "attributes": Object {},
      "children": Array [
        TPhrasing {
          "attributes": Object {},
          "children": Array [
            TText {
              "attributes": Object {},
              "children": Array [],
              "data": "Line2",
              "hasWhiteSpaceCollapsingEnabled": true,
              "isAnchor": false,
              "parentStyles": TStyles {
                "nativeBlockFlow": Object {},
                "nativeBlockRet": Object {},
                "nativeTextFlow": Object {},
                "nativeTextRet": Object {},
                "webTextFlow": Object {},
              },
              "styles": TStyles {
                "nativeBlockFlow": Object {},
                "nativeBlockRet": Object {},
                "nativeTextFlow": Object {},
                "nativeTextRet": Object {},
                "webTextFlow": Object {},
              },
              "tagName": null,
              "type": "text",
            },
            TText {
              "attributes": Object {},
              "children": Array [],
              "data": " Line3",
              "hasWhiteSpaceCollapsingEnabled": true,
              "isAnchor": false,
              "parentStyles": TStyles {
                "nativeBlockFlow": Object {},
                "nativeBlockRet": Object {},
                "nativeTextFlow": Object {},
                "nativeTextRet": Object {},
                "webTextFlow": Object {},
              },
              "styles": TStyles {
                "nativeBlockFlow": Object {},
                "nativeBlockRet": Object {},
                "nativeTextFlow": Object {},
                "nativeTextRet": Object {},
                "webTextFlow": Object {},
              },
              "tagName": "span",
              "type": "text",
            },
          ],
          "hasWhiteSpaceCollapsingEnabled": true,
          "isAnchor": false,
          "parentStyles": TStyles {
            "nativeBlockFlow": Object {},
            "nativeBlockRet": Object {},
            "nativeTextFlow": Object {},
            "nativeTextRet": Object {},
            "webTextFlow": Object {},
          },
          "styles": TStyles {
            "nativeBlockFlow": Object {},
            "nativeBlockRet": Object {},
            "nativeTextFlow": Object {},
            "nativeTextRet": Object {},
            "webTextFlow": Object {},
          },
          "tagName": null,
          "type": "phrasing",
        },
        TBlock {
          "attributes": Object {
            "src": "https://domain.com/logo.jpg",
          },
          "children": Array [],
          "hasWhiteSpaceCollapsingEnabled": true,
          "isAnchor": false,
          "parentStyles": TStyles {
            "nativeBlockFlow": Object {},
            "nativeBlockRet": Object {},
            "nativeTextFlow": Object {},
            "nativeTextRet": Object {},
            "webTextFlow": Object {},
          },
          "styles": TStyles {
            "nativeBlockFlow": Object {},
            "nativeBlockRet": Object {},
            "nativeTextFlow": Object {},
            "nativeTextRet": Object {},
            "webTextFlow": Object {},
          },
          "tagName": "img",
          "type": "block",
        },
      ],
      "hasWhiteSpaceCollapsingEnabled": true,
      "isAnchor": false,
      "parentStyles": null,
      "styles": TStyles {
        "nativeBlockFlow": Object {},
        "nativeBlockRet": Object {},
        "nativeTextFlow": Object {},
        "nativeTextRet": Object {},
        "webTextFlow": Object {},
      },
      "tagName": "strong",
      "type": "block",
    },
  ],
  "hasWhiteSpaceCollapsingEnabled": true,
  "isAnchor": false,
  "parentStyles": null,
  "styles": TStyles {
    "nativeBlockFlow": Object {},
    "nativeBlockRet": Object {},
    "nativeTextFlow": Object {},
    "nativeTextRet": Object {},
    "webTextFlow": Object {},
  },
  "tagName": "span",
  "type": "block",
}
`;
